File: project/server/src/Domain/Dto/Auth/UsersDto.cs
namespace server.src.Domain.Dto.Auth;

/// <summary>
/// Represents a summarized user record for listing purposes.
/// </summary>
/// <param name="Id">The unique identifier of the user.</param>
/// <param name="FirstName">The user's first name.</param>
/// <param name="LastName">The user's last name.</param>
/// <param name="Email">The user's email address.</param>
/// <param name="UserName">The user's username.</param>
/// <param name="PhoneNumber">The user's phone number.</param>
/// <param name="MobilePhoneNumber">The user's mobile phone number.</param>
public record ListItemUserDto(
    string Id,
    string FirstName,
    string LastName,
    string Email,
    string UserName,
    string PhoneNumber,
    string MobilePhoneNumber
);

/// <summary>
/// Represents detailed information about a user.
/// </summary>
/// <param name="Id">The unique identifier of the user.</param>
/// <param name="FirstName">The user's first name.</param>
/// <param name="LastName">The user's last name.</param>
/// <param name="Email">The user's email address.</param>
/// <param name="EmailConfirmed">Indicates whether the email is confirmed.</param>
/// <param name="UserName">The user's username.</param>
/// <param name="LockoutEnabled">Indicates whether lockout is enabled for the user.</param>
/// <param name="LockoutEnd">The end date of the lockout period, if applicable.</param>
/// <param name="InitialPassword">The user's initial password (if applicable).</param>
/// <param name="Address">The user's physical address.</param>
/// <param name="ZipCode">The user's ZIP code.</param>
/// <param name="City">The city of the user's address.</param>
/// <param name="State">The state of the user's address.</param>
/// <param name="Country">The country of the user's address.</param>
/// <param name="PhoneNumber">The user's phone number.</param>
/// <param name="PhoneNumberConfirmed">Indicates whether the phone number is confirmed.</param>
/// <param name="MobilePhoneNumber">The user's mobile phone number.</param>
/// <param name="MobilePhoneNumberConfirmed">Indicates whether the mobile phone number is confirmed.</param>
/// <param name="Bio">A brief biography of the user.</param>
/// <param name="DateOfBirth">The user's date of birth.</param>
/// <param name="IsActive">Indicates whether the user is active.</param>
/// <param name="Roles">The list of roles assigned to the user.</param>
public record ItemUserDto(
    string Id,
    string FirstName,
    string LastName,
    string Email,
    bool EmailConfirmed,
    string UserName,
    bool LockoutEnabled,
    string LockoutEnd,
    string InitialPassword,
    string Address,
    string ZipCode,
    string City,
    string State,
    string Country,
    string PhoneNumber,
    bool PhoneNumberConfirmed,
    string MobilePhoneNumber,
    bool MobilePhoneNumberConfirmed,
    string Bio,
    string DateOfBirth,
    bool IsActive,
    List<string> Roles
);

/// <summary>
/// Represents the data required to create or update a user.
/// </summary>
/// <param name="FirstName">The user's first name.</param>
/// <param name="LastName">The user's last name.</param>
/// <param name="Email">The user's email address.</param>
/// <param name="UserName">The user's username.</param>
/// <param name="Password">The user's password.</param>
/// <param name="Address">The user's physical address.</param>
/// <param name="ZipCode">The ZIP code of the user's address.</param>
/// <param name="City">The city of the user's address.</param>
/// <param name="State">The state of the user's address.</param>
/// <param name="Country">The country of the user's address.</param>
/// <param name="PhoneNumber">The user's phone number.</param>
/// <param name="MobilePhoneNumber">The user's mobile phone number.</param>
/// <param name="Bio">A brief biography of the user.</param>
/// <param name="DateOfBirth">The user's date of birth.</param>
public record UserDto(
    string FirstName,
    string LastName,
    string Email,
    string UserName,
    string Password,
    string Address,
    string ZipCode,
    string City,
    string State,
    string Country,
    string PhoneNumber,
    string MobilePhoneNumber,
    string Bio,
    DateTime DateOfBirth
);

/// <summary>
/// Represents the credentials required for a user login request.
/// </summary>
/// <param name="Username">The username of the user.</param>
/// <param name="Password">The password of the user.</param>
public record UserLoginDto(string Username, string Password);

/// <summary>
/// Represents an authenticated user with their username and access token.
/// </summary>
/// <param name="UserName">The username of the authenticated user.</param>
/// <param name="Token">The access token for the authenticated session.</param>
public record AuthenticatedUserDto(string UserName, string Token);

/// <summary>
/// Represents the data required to request a password reset.
/// </summary>
/// <param name="Email">The email address of the user requesting a password reset.</param>
public record ForgotPasswordDto(string Email);

/// <summary>
/// Represents the data required to reset a user's password.
/// </summary>
/// <param name="Email">The user's email address.</param>
/// <param name="Token">The token required to authorize the password reset.</param>
/// <param name="NewPassword">The new password for the user.</param>
public record ResetPasswordDto(string Email, string Token, string NewPassword);

/// <summary>
/// Represents the data required to enable two-factor authentication for a user.
/// </summary>
/// <param name="UserId">The unique identifier of the user.</param>
public record Enable2FADto(string UserId);

/// <summary>
/// Represents the data required to verify two-factor authentication for a user.
/// </summary>
/// <param name="UserId">The unique identifier of the user.</param>
/// <param name="Token">The verification token for two-factor authentication.</param>
public record Verify2FADto(string UserId, string Token);

File: project/client/src/models/auth/usersDto.ts
/**
 * Represents a minimal representation of a user with basic information.
 * 
 * This interface is used for listing users, including their name, email, username, and phone numbers.
 */
export interface ListItemUserDto {
    /**
     * The unique identifier for the user.
     */
    id: string;

    /**
     * The user's first name.
     */
    firstName: string;

    /**
     * The user's last name.
     */
    lastName: string;

    /**
     * The user's email address.
     */
    email: string;

    /**
     * The user's username.
     */
    userName: string;

    /**
     * The user's phone number.
     */
    phoneNumber: string;

    /**
     * The user's mobile phone number.
     */
    mobilePhoneNumber: string;
}

/**
 * Represents detailed information about a user, including personal and contact information, as well as account status.
 * 
 * This interface is used for full user data, including security details such as lockout status and roles.
 */
export interface ItemUserDto {
    /**
     * The unique identifier for the user.
     */
    id: string;
  
    /**
     * The user's first name.
     */
    firstName: string;
  
    /**
     * The user's last name.
     */
    lastName: string;
  
    /**
     * The user's email address.
     */
    email: string;
  
    /**
     * Whether the user's email has been confirmed.
     */
    emailConfirmed: boolean;
  
    /**
     * The user's username.
     */
    userName: string;
  
    /**
     * Whether the user is locked out of their account.
     */
    lockoutEnabled: boolean;
  
    /**
     * The date and time the user's lockout ends, if applicable.
     */
    lockoutEnd: string;
  
    /**
     * The user's initial password (used for user creation).
     */
    initialPassword: string;
  
    /**
     * The user's home address.
     */
    address: string;
  
    /**
     * The user's zip code.
     */
    zipCode: string;
  
    /**
     * The user's city.
     */
    city: string;
  
    /**
     * The user's state.
     */
    state: string;
  
    /**
     * The user's country.
     */
    country: string;
  
    /**
     * The user's phone number.
     */
    phoneNumber: string;
  
    /**
     * Whether the user's phone number is confirmed.
     */
    phoneNumberConfirmed: boolean;
  
    /**
     * The user's mobile phone number.
     */
    mobilePhoneNumber: string;
  
    /**
     * Whether the user's mobile phone number is confirmed.
     */
    mobilePhoneNumberConfirmed: boolean;
  
    /**
     * A short biography or description of the user.
     */
    bio: string;
  
    /**
     * The user's date of birth.
     */
    dateOfBirth: string;
  
    /**
     * Whether the user's account is active.
     */
    isActive: boolean;
}
  
/**
 * Represents the data used to create or update a user account.
 * 
 * This interface is used for user registration or profile creation, including all relevant personal information.
 */
export interface UserDto {
    /**
     * The user's first name.
     */
    firstName: string;
  
    /**
     * The user's last name.
     */
    lastName: string;
  
    /**
     * The user's email address.
     */
    email: string;
  
    /**
     * The user's username.
     */
    userName: string;
  
    /**
     * The user's password.
     */
    password: string;
  
    /**
     * The user's home address.
     */
    address: string;
  
    /**
     * The user's zip code.
     */
    zipCode: string;
  
    /**
     * The user's city.
     */
    city: string;
  
    /**
     * The user's state.
     */
    state: string;
  
    /**
     * The user's country.
     */
    country: string;
  
    /**
     * The user's phone number.
     */
    phoneNumber: string;
  
    /**
     * The user's mobile phone number.
     */
    mobilePhoneNumber: string;
  
    /**
     * A short biography or description of the user.
     */
    bio: string;
  
    /**
     * The user's date of birth.
     */
    dateOfBirth: Date;
}
  
/**
 * Represents the data used by a user to log in.
 * 
 * This interface includes the user's credentials (username and password).
 */
export interface UserLoginDto {
    /**
     * The user's username.
     */
    username: string;
  
    /**
     * The user's password.
     */
    password: string;
  }
  
/**
 * Represents a successfully authenticated user.
 * 
 * This interface contains the username and a token used for authentication in future requests.
 */
export interface AuthenticatedUserDto {
    /**
     * The username of the authenticated user.
     */
    userName: string;
  
    /**
     * A token used to authenticate the user in future requests.
     */
    token: string;
}
  
  /**
   * Represents the data used to request a password reset.
   * 
   * This interface includes the user's email address.
   */
  export interface ForgotPasswordDto {
    /**
     * The email address of the user requesting the password reset.
     */
    email: string;
  }
  
/**
 * Represents the data used to reset a password.
 * 
 * This interface includes the user's email, the reset token, and the new password.
 */
export interface ResetPasswordDto {
    /**
     * The email address of the user resetting the password.
     */
    email: string;
  
    /**
     * The token provided for the password reset process.
     */
    token: string;
  
    /**
     * The new password for the user.
     */
    newPassword: string;
}
  
/**
 * Represents the data used to enable Two-Factor Authentication (2FA) for a user.
 * 
 * This interface contains the user's ID.
 */
export interface Enable2FADto {
    /**
     * The unique identifier of the user enabling 2FA.
     */
    userId: string;
}
  
/**
 * Represents the data used to verify Two-Factor Authentication (2FA) for a user.
 * 
 * This interface includes the user's ID and the verification token.
 */
export interface Verify2FADto {
    /**
     * The unique identifier of the user verifying 2FA.
     */
    userId: string;
  
    /**
     * The token used to verify the 2FA process.
     */
    token: string;
}  